pki: ipsec: T3642: Add new PKI configuration and IPSec integration

sarthurdev opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a commend and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Adds PKI library and configuration to VyOS, with ability to generate and store certificates, CA certificates, keys etc. Also includes initial integration for IPSec.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe): 

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T3642

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
pki, ipsec

## Proposed changes
<!--- Describe your changes in detail -->
Adds PKI configuration to VyOS, ability to generate and store certificates, CA certificates, keys etc. Uses python3-cryptography as the new backend for the `vyos.pki` module.

Includes x509 integration for IPSec, smoketest, XML templates for easier integration with other nodes.

Adds the following config nodes:
```
set pki ca <name> certificate <certificate pem data>
set pki ca <name> private key <private key pem data>
set pki ca <name> private password-protected # If node set, private key treated as encrypted, generation tasks will prompt for password entry
set pki ca <name> crl <crl pem data> # Multiple CRL values allowed
set pki certificate <name> certificate <certificate pem data>
set pki certificate <name> private key <private key pem data>
set pki certificate <name> private password-protected # If node set, private key treated as encrypted
set pki certificate <name> revoke # If node set, certificate will be added to CRL when CRL generation task is executed on it's CA
set pki dh <name> parameters <dh params pem data>
set pki key-pair <name> public key <public key pem data> # Allow for public keys to be added, as private key may not be present on this host
set pki key-pair <name> private key <private key pem data>
set pki key-pair <name> private password-protected # If node set, private key treated as encrypted
set pki openvpn shared-secret <name> key <tls-auth key data>
set pki openvpn shared-secret <name> version '1' # not required, just for future proofing if key format changes
set pki x509 default country/state/locality/organization <value> # Allows for default values to be set for certificate generation
```

Op-mode:
Note: All generation tasks allow for the suffix `install <conf name>` to be added, if present the generation command will output conf-mode set commands instead of the raw PEM data.
```
generate pki ca [install <name>] # Generates a CA certificate and private key
generate pki certificate [install <name>] # Generate certificate request and private key
generate pki certificate self-sign [install <name>] # Generates a self-signed certificate and private key
generate pki certificate sign <ca name> [install <name>] # Generates a certificate and private key, signed by specified CA
generate pki crl <ca name> [install <name> # Generates a CRL for all revoke marked certificates for specified CA
generate pki dh [install <name>] # Generates Diffie-Hellman parameters
generate pki key-pair [install <name>] # Generates a private key and outputs public key
generate pki openvpn shared-secret [install <name>] # Generates an OpenVPN static key
generate pki ssh-key [install <name>] # Generates an OpenSSH private key, install command will show conf-mode commands for the logged in user
generate pki wireguard key-pair [install <name>] # Generates a wireguard key-pair
generate pki wireguard psk [install <name>] # Generates a wireguard pre shared key

show pki ca [name] # Show CA certificate(s)
show pki certificate [name] # Show certificate(s) and relationships
show pki crl [ca name] # Show CRLs for a CA and which certificates they revoke (if present)
```

If merged, future PRs will contain further migrations of nodes currently using file-based certificates and integrate with PKI config.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
* Tested and ensured all generation commands output valid data and can be import with supplied conf-mode commands.
* Tested IPSec migration and PKI implementation using VM snapshot with old version and upgrading with new ISO. CA, certs and keys were migrated successfully.
* Added smoketest for conf-mode verifications

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sarthurdev committed:
pki: T3642: New PKI config and management

sarthurdev committed:
pki: ipsec: T3642: Migrate IPSec to use PKI configuration

c-po commented on interface-definitions/include/pki/certificate-key.xml.i:
```diff
@@ -0,0 +1,23 @@
+<!-- include start from pki/certificate-key.xml.i -->
+<leafNode name="certificate">
```
Any reason to not include `interface-definitions/include/pki/certificate.xml.i` here? Seems to be a perfekt fit.

c-po commented on interface-definitions/include/pki/certificate-key.xml.i:
```diff
@@ -0,0 +1,23 @@
+    </completionHelp>
+  </properties>
+</leafNode>
+<leafNode name="private-key-passphrase">
```
Can we rename this to simply be `passphrase`?

c-po commented on interface-definitions/include/pki/dh-parameters.xml.i:
```diff
@@ -0,0 +1,14 @@
+<!-- include start from pki/dh-parameters.xml.i -->
```
Where is this file referenced? If there are parameters, there should be a completion helper?

sarthurdev replied:
This file exists for future use, for example here: https://github.com/vyos/vyos-1x/blob/current/interface-definitions/interfaces-openvpn.xml.in#L683

c-po replied:
Please do not add includes for future use - they will probably not be picked up until you really know it is there. Just send it in a future PR when needed.

c-po commented on interface-definitions/include/pki/openvpn_tls-auth.xml.i:
```diff
@@ -0,0 +1,14 @@
+<!-- include start from pki/openvpn_tls-auth.xml.i -->
```
Well I am unsure if there is a need for an "auth key" CLI node - it is the same as an OpenVPN site2site key if I remember correctly. We should find a common name for the CLI node.

c-po commented on interface-definitions/pki.xml.in:
```diff
@@ -0,0 +1,203 @@
+          </node>
+        </children>
+      </tagNode>
+      <node name="openvpn">
```
What about an OpenVPN Site2Site key aka `shared-secret-key-file` CLI option?

sarthurdev replied:
Do you think its best if I just rename the tls-auth node to a common name of `static-key`? Then in a future PR where OpenVPN is integrated we can reference that single node for both `shared-secret-key` and `tls-auth`.

c-po replied:
I'd simply call it `shared-secret`, as that is what OpenVPN calls it internally:

```
--genkey        : Generate a random key to be used as a shared secret,
                  for use with the --secret option.
```

sarthurdev replied:
Changed

c-po commented:
Besides my requested changes it would be nice to have an IPSec configuration in https://github.com/vyos/vyos-1x/tree/current/smoketest/configs which makes use of the old PKI infrastructure so the migration script is also CI/CD validated on each build. SImply add your sanitized config here. If key's need to be created in advance to pass the test or reference in your configuration, piease take a look at:

https://github.com/vyos/vyos-build/blob/current/scripts/check-qemu-install#L360-L387


Here I added some code which will generate a set of new SSL certificates on each test invokation

sarthurdev commented:
> 
> 
> Besides my requested changes it would be nice to have an IPSec configuration in https://github.com/vyos/vyos-1x/tree/current/smoketest/configs which makes use of the old PKI infrastructure so the migration script is also CI/CD validated on each build. SImply add your sanitized config here. If key's need to be created in advance to pass the test or reference in your configuration, piease take a look at:
> 
> https://github.com/vyos/vyos-build/blob/current/scripts/check-qemu-install#L360-L387
> 
> Here I added some code which will generate a set of new SSL certificates on each test invokation

Added a smoketest config using the test files. Running the test now.

sarthurdev committed:
pki: ipsec: T3642: Update migration script to account for file permission issues
